File: src/config/64_to_32.pl
#
# $Id$
#
#
#
# perl script to do transliteration from "double" values to "8wrap" values
#
# The script is envoked with the command:
#    perl 64_to_32.pl file1.f [file2.f ...]
#
# This script uses $NWCHEM_TOP/src/config/data.64_to_32 which is a datafile 
# that containes the specific transliterations for "double" to "8wrap" in 
# simple ascii format.  If you need to add a conversion see the comments in 
# the data file.
# Tthis conversion is needed for using 32-bit integer libraries (blas,lapack
# and scalapack) when the rest of the code uses 64-bit integers
# 8wrap files convert from 64-bit to 32-bit
# 
# Written:  3/14/97
# By:       Ricky A. Kendall
#           High Performance Computational Chemistry Group
#           Theory Modeling and Simulation Program
#           Environmental Molecular Sciences Laboratory
#           Pacific Northwest National Laboratory
#           P.O. Box 999
#           Richland, WA 99352-0999
#

sub copy_case() {   # Take case from "from" and apply it to "to" and return that new string
    my ($from, $to) = @_;
    my ($lf, $lt) = map length, @_;
    if ($lt < $lf) { $from = substr $from, 0, $lt }
    else { $from .= substr $to, $lf }
    return uc $to | ($from ^ uc $from);
}                   
 
$debug = 0;
@from = ();
@to   = ();
use File::stat;
use File::Basename;
$data_path = dirname(__FILE__);
if($debug) {print "{$data_path} \n";}

#$data_path = $data_path . "/src/config/data.64_to_32";
$data_path = $data_path . "/data.64_to_32";
if($debug) {print "{$data_path} \n";}
open (DATA,$data_path) || die " unable to open: $data_path \n";
while (<DATA>)
{
    if (/^[^\#]/) {
	if($debug) {print $_;}
	@tokens = split(' ');      
	$num_tokens = @tokens ;
	if($debug){print "tokens: @tokens $#tokens $num_tokens \n";}
	push(@from,$tokens[0]);
	push(@to,  $tokens[1]);
    }
}
close (DATA);
if($debug){
    print "from array @from \n";
    print "to   array @to \n";
}
$num_from = @from;
$num_to   = @to;
if ($num_from != $num_to) {
    print "To and From token count not identical\n";
    print "Number of From tokens: $num_from\n";
    print "number of To   tokens: $num_to\n";
    die "Fatal dbl2sngl error\n";
}
else
{
    if ($debug){
	print "Number of From tokens: $num_from\n";
	print "number of To   tokens: $num_to\n";
    }
    $num_compare = $num_from;
}
if ($debug) { print "arguments: @ARGV\n";}
foreach my $file (@ARGV){
    if ($debug){print "file        : $file\n";}
    my $orgfile = $file;
    $filebak = $file . ".$$" ;
    if ($debug){print "backup file : $filebak\n";}
    rename($file,$filebak);
    $file = '>' . $file;
    my $my_pid = $pid;
    if (open(FILETOFIX,$filebak)) {
#	print " open OK\n";
    }else{
	print "Exiting: Could not open file: $filebak\n";
	kill 15, -$my_pid;
    }
    open(FIXEDFILE,$file) || die "Could not open file: $file\n";
    while (<FILETOFIX>) {
	if ( /^c/ || /^C/ || /^\*/ || /^$/){
	    print FIXEDFILE $_;
	}
	else	{
	    for ($compare = 0; $compare < $num_compare ; $compare++)
	    {
		if (/$from[$compare]/i){

		    $froom = index(uc $_,uc $from[$compare]); # Find where the string starts ie the "from"
		    $toot = &copy_case(substr($_,$froom),$to[$compare]) ; # Generate a "to" that matches the case

		    if (/^[ ]{5}[^\s]/) {
			s/([ ]{5}.)$from[$compare](\W{1})/$1${toot}$2/gi ;
		    }
#this takes care of the tab chars in C
		    if (/^[ \t]/){
			s/(\W{1})$from[$compare](\W{1})/$1${toot}$2/gi ;
		    }
#this takes care of declarations in C
		    if (/^[ \S]/){
			s/(\W{1})$from[$compare](\W{1})/$1${toot}$2/gi ;
		    }
		    if (/^[ \d]/){
			s/(\W{1})$from[$compare](\W{1})/$1${toot}$2/gi ;
		    }
		}
	    }
	    print FIXEDFILE $_;
	}
    }
    close(FIXEDFILE);
    close(FILETOFIX);
    unlink($filebak);
    my $filesize = stat("$orgfile")->size;
    if($filesize eq "0") {
	print("Exiting: converted file $orgfile has size $filesize\n");
	kill 15, -$my_pid;
    }
}


	
	

File: src/config/64_to_32.py
#!/usr/bin/env python3
"""
Python script to do transliteration from "double" values to "8wrap" values
Usage: python 64_to_32.py file1.f [file2.f ...]

Original Perl script:
    Written:  3/14/97
    By:       Ricky A. Kendall

Converted from Perl to Python using Claude (claude-4-5-opus-4-5-20251101-v1)
Optimizations applied:
    - Early exits to skip unnecessary processing
    - Reduced redundant regex passes
    - Combined line-type checks
    - Precompiled regex patterns
"""

import os
import sys
import re
import subprocess # nosec B404

debug = False


def copy_case(from_str, to_str):
    """Take case from 'from_str' and apply it to 'to_str' and return that new string"""
    result = []
    for i, char in enumerate(to_str):
        if i < len(from_str):
            if from_str[i].isupper():
                result.append(char.upper())
            elif from_str[i].islower():
                result.append(char.lower())
            else:
                result.append(char)
        else:
            result.append(char)
    return ''.join(result)


class ConversionPattern:
    """Holds precompiled regex patterns for a single from/to conversion pair"""
    
    def __init__(self, from_str, to_str):
        self.from_str = from_str
        self.to_str = to_str
        self.from_str_upper = from_str.upper()
        
        # Precompile all regex patterns
        escaped = re.escape(from_str)
        
        # Pattern for finding the from_str (case insensitive)
        self.search_pattern = re.compile(escaped, re.IGNORECASE)
        
        # Pattern for Fortran continuation lines (5 spaces + non-space)
        self.fortran_pattern = re.compile(
            r'([ ]{5}.)' + escaped + r'(\W)',
            re.IGNORECASE
        )
        
        # Pattern for general substitution (word boundaries)
        self.general_pattern = re.compile(
            r'(\W)' + escaped + r'(\W)',
            re.IGNORECASE
        )


def load_data_file():
    """Load the from/to conversion pairs from the data file and precompile patterns"""
    patterns = []
    
    data_path = os.path.dirname(os.path.abspath(__file__))
    data_path = os.path.join(data_path, "data.64_to_32")
    
    if debug:
        print(f"Data path: {data_path}")

    try:
        with open(data_path, 'r') as data_file:
            for line in data_file:
                if line and not line.startswith('#'):
                    tokens = line.split()
                    if len(tokens) >= 2:
                        # For 64_to_32: tokens[0] is "from", tokens[1] is "to" [1]
                        patterns.append(ConversionPattern(tokens[0], tokens[1]))
    except IOError:
        sys.exit(f"Unable to open: {data_path}")
    
    if debug:
        print(f"Loaded {len(patterns)} conversion patterns")
    
    return patterns


# Precompile line-type detection patterns (used for every line)
FORTRAN_CONTINUATION = re.compile(r'^[ ]{5}[^\s]')


def process_line(line, patterns, line_upper):
    """
    Process a single line with all conversion patterns.
    Uses early exits and optimized regex passes.
    """
    # Early exit: Skip comment lines (Fortran style) or empty lines [1]
    if not line or line[0] in ('c', 'C', '*') or line.strip() == '':
        return line
    
    # Early exit: Quick check if any pattern might match using uppercase comparison
    # This avoids regex overhead for lines with no potential matches
    has_potential_match = False
    for conv in patterns:
        if conv.from_str_upper in line_upper:
            has_potential_match = True
            break
    
    if not has_potential_match:
        return line
    
    # Determine line type once (not for each pattern) [1]
    is_fortran_continuation = line.startswith('     ') and len(line) > 5 and not line[5].isspace()
    is_general_line = len(line) > 0 and line[0] in ' \t'
    
    # Process each conversion pattern
    for conv in patterns:
        # Early exit: Skip if pattern not in line (case-insensitive quick check)
        if conv.from_str_upper not in line_upper:
            continue
        
        match = conv.search_pattern.search(line)
        if not match:
            continue
        
        # Found a match - compute replacement
        froom = match.start()
        toot = copy_case(
            line[froom:froom + len(conv.from_str)],
            conv.to_str
        )
        
        # Apply appropriate substitution based on line type (single pass) [1]
        if is_fortran_continuation:
            line = conv.fortran_pattern.sub(
                lambda m: m.group(1) + toot + m.group(2),
                line
            )
        
        if is_general_line:
            line = conv.general_pattern.sub(
                lambda m: m.group(1) + toot + m.group(2),
                line
            )
        
        # Update line_upper after substitution for subsequent patterns
        line_upper = line.upper()
    
    return line


def process_file(file_path, patterns):
    """Process a single file and perform transliteration"""
    pid = os.getpid()
    orgfile = file_path
    filebak = f"{file_path}.{pid}"
    
    if debug:
        print(f"Processing: {file_path}")
        print(f"Backup file: {filebak}")
    
    # Runs the Linux 'file' command
    
    result = subprocess.run(['file', file_path], shell=False, capture_output=True, text=True) # nosec

    # DOS files output contains "with CRLF line terminators"
    if "CRLF" in result.stdout:
        print(f"ERROR: DOS file with CRLF {file_path}", file=sys.stderr)
        sys.exit(123)  
    
    # Rename original to backup [1]
    os.rename(file_path, filebak)
    
    try:
        with open(filebak, 'r') as fh_in:
            with open(orgfile, 'w') as fh_out:
                for line in fh_in:
                    line_upper = line.upper()
                    processed_line = process_line(line, patterns, line_upper)
                    fh_out.write(processed_line)
    except IOError as e:
        sys.exit(f"Can't open file: {e}")
    
    # Remove backup [1]
    os.unlink(filebak)


def main():
    if debug:
        print(f"Arguments: {sys.argv[1:]}")
    
    # Load patterns once (precompiled)
    patterns = load_data_file()
    
    if len(patterns) == 0:
        sys.exit("Fatal dbl2sngl error: No conversion patterns loaded")
    
    files = sys.argv[1:]
    
    if len(files) == 0:
        print("Usage: python 64_to_32.py file1.f [file2.f ...]")
        sys.exit(1)
    
    # Process each file sequentially
    for file_path in files:
        process_file(file_path, patterns)


if __name__ == "__main__":
    main()
